### PR TITLE
Refactor asset create/update/delete

### PR DESCRIPTION
**Pull request by danlamanna**
This is a general refactor of asset create/update/delete code. This PR doesn't alter any existing functionality, it just restructures the internal API.

@AlmightyYakob cc'ing you because this slightly touches asset path code

### CODE REVIEW

**Review by AlmightyYakob — CHANGES_REQUESTED** [review]
This looks great! Just a few questions/comments

**Review by mvandenburgh — COMMENTED** [review]
Same questions as Jake but otherwise this looks good to me :+1:

File: dandiapi/api/services/asset/__init__.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        asset = Asset(
+            path=path,
+            blob=asset_blob,
+            embargoed_blob=embargoed_asset_blob,
+            zarr=zarr_archive,
+            metadata=metadata,
+            status=Asset.Status.PENDING,
+        )
+        asset.full_clean(validate_constraints=False)
+        asset.save()
```
[the remark this answers is not in the document]
Comment (danlamanna) [reply dandiapi/api/services/asset/__init__.py]: No reason, probably just a vestige of some copy/paste.

File: dandiapi/api/services/asset/metadata.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    if asset.is_blob:
+        blob = asset.blob
+    elif asset.is_embargoed_blob:
+        blob = asset.embargoed_blob
+    else:
+        # TODO: assert? zarr?
+        return
+
+    if blob.sha256 is None:
+        return
```
[the remark this answers is not in the document]
Comment (danlamanna) [reply dandiapi/api/services/asset/metadata.py:24]: I don't disagree, but I'm just moving this function so I would rather not change behavior in this PR.

File: dandiapi/api/services/asset/__init__.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        return asset, False
+
+    # Verify we aren't changing path to the same value as an existing asset
+    if version.assets.filter(path=path).exclude(asset_id=asset.asset_id).exists():
+        raise AssetAlreadyExists()
+
+    with transaction.atomic():
+        # make sure to check remove_asset_from_version in order to handle any
+        # internal details of removing assets.
+        version.assets.remove(asset)
```
[the remark this answers is not in the document]
Comment (danlamanna) [reply dandiapi/api/services/asset/__init__.py]: It should, I fixed it.

**Review by AlmightyYakob — APPROVED** [review]

**Pull request merged**